plugins/analogix/fu-analogix-device.c

```diff
@@ -0,0 +1,556 @@
+/*
+ * Copyright (C) 2021 Xiaotian Cui <xtcui@analogixsemi.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+#include "config.h"
+
+#include <string.h>
+
+#include "fu-chunk.h"
+
+#include "fu-analogix-common.h"
+#include "fu-analogix-device.h"
+#include "fu-analogix-firmware.h"
+
+
+struct _FuAnalogixDevice {
+	FuUsbDevice	parent_instance;
+	guint8		iface_idx;		/* bInterfaceNumber */
+	guint8		ep_num;			/* bEndpointAddress */
+	guint16		chunk_len;		/* wMaxPacketSize */
+	guint16		vid;
+	guint16		pid;
+	guint16		rev;
+	guint16		custom_version;
+	guint16		fw_version;
+};
+
+G_DEFINE_TYPE (FuAnalogixDevice, fu_analogix_device, FU_TYPE_USB_DEVICE)
+
+static gboolean
+fu_analogix_device_send (FuAnalogixDevice *self,
+			AnxBbRqtCode reqcode,
+			guint16 val0code,
+			guint16 index,
+			guint8 *in,
+			gsize in_len,
+			GError **error)
+{
+	gsize actual_len = 0;
+
+	/* check size */
+	if (in_len > 64) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "input buffer too large");
+		return FALSE;
+	}
+	g_return_val_if_fail (in != NULL, FALSE);
+
+	/* send data to device */
+	if (!g_usb_device_control_transfer (
+		    fu_usb_device_get_dev (FU_USB_DEVICE (self)),
+		    G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+		    G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+		    G_USB_DEVICE_RECIPIENT_DEVICE,
+		    reqcode, /* request */
+		    val0code, /* value */
+		    index, /* index */
+		    in, /* data */
+		    in_len, /* length */
+		    &actual_len, /* actual length */
+		    (guint) ANX_BB_TRANSACTION_TIMEOUT,
+		    NULL, error)) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "send data error");
+		return FALSE;
+	}
+	if (actual_len != in_len) {
+		g_prefix_error (error, "receive data error count: ");
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "send data length is incorrect");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_analogix_device_receive (FuAnalogixDevice *self,
+			    AnxBbRqtCode reqcode,
+			    guint16 val0code,
+			    guint16 index,
+			    guint8 *out,
+			    gsize out_len,
+			    GError **error)
+{
+	gsize actual_len = 0;
+
+	/* check size */
+	if (out_len > 64) {

```

Comment:
everywhere this function is called this is a hardcoded value.  I think this should be a `g_return_val_if_false` indicating programming error if anything.